File: Strings/Algorithms/PrefixTrie.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_end_of_word = False

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        current = self.root
        for char in word:
            if char not in current.children:
                current.children[char] = TrieNode()
            current = current.children[char]
        current.is_end_of_word = True

    def search(self, word):
        current = self.root
        for char in word:
            if char not in current.children:
                return False
            current = current.children[char]
        return current.is_end_of_word

    def startsWith(self, prefix):
        current = self.root
        for char in prefix:
            if char not in current.children:
                return False
            current = current.children[char]
        return True

prefix_trie = Trie()

prefix_trie.insert("apple")
print(f"Search 'apple': {prefix_trie.search('apple')}")      # Expected: True
print(f"Search 'app': {prefix_trie.search('app')}")          # Expected: False
print(f"Starts with 'app': {prefix_trie.startsWith('app')}")  # Expected: True

prefix_trie.insert("app")
print(f"Search 'app' after insert: {prefix_trie.search('app')}")

File: Strings/Algorithms/PrefixTrie.cpp
#include <bits/stdc++.h>

using namespace std;
typedef long long ll;
typedef long double ld;
const ll INF = numeric_limits<ll>::max();

#define all(x) x.begin(), x.end()
#define rall(x) x.rbegin(), x.rend()
#define MOD 998244353
#define pi acos(-1)
#define gcd __gcd
#define lcm(a, b) ((a) * (b) / gcd(a, b))
#define precision(n) cout << fixed << setprecision(n)
#define debug(arr) cerr << #arr << " = "; for (auto &x : arr) cerr << x << " "; cerr << endl;
#define fastio ios::sync_with_stdio(0); cin.tie(0); cout.tie(0);

struct Node{
    int children[26];
    bool flag = false; 

    Node(){
        fill(begin(children), end(children), -1);
    }
};

struct Trie {
    vector<Node> Tree;
    Trie() {
        Tree.emplace_back();
    }
    
    void insert(string word) {
        int v = 0;
        for(int i = 0; i < word.size(); i++){
            int code = word[i] - 'a';
            if(Tree[v].children[code] == -1){
                Tree[v].children[code] = Tree.size();
                Tree.emplace_back();
            }
            v = Tree[v].children[code];   
        }
        Tree[v].flag = true;
    }
    
    bool search(string word) {
      int v = 0;
      for(int i = 0; i < word.size(); i++){
        int code = word[i] - 'a';
        if(Tree[v].children[code] == -1) return false;
        v = Tree[v].children[code];
      }
      return Tree[v].flag;
    }
    
    bool startsWith(string prefix) {
      int v = 0;
      for(int i = 0; i < prefix.size(); i++){
        int code = prefix[i] - 'a';
        if(Tree[v].children[code] == -1) return false;
        v = Tree[v].children[code];
      }
      return true;
    }
};

int main(){
    fastio;
    
    Trie PrefixTrie;
    PrefixTrie.insert("apple");
    cout << PrefixTrie.search("apple") << endl;
    cout << PrefixTrie.search("app") << endl;
    cout << PrefixTrie.startsWith("app") << endl;
    PrefixTrie.insert("app");
    cout << PrefixTrie.search("app") << endl;
}